source/_posts/vue-design-patterns.md

```diff
@@ -0,0 +1,982 @@
+
+That problem is known as **props drilling** (an anti-pattern). Fortunately, Vue has the solution build-in with its `provide()` and `inject()` features, better known as dependency injection.
+
+{% img "3-provide-inject.png" "Using the provide/inject feature from Vue" "lazy" %}
+
+As we can see in the image above, **Provide** gets the data and is responsible for passing it when it is needed to be **Injected** in all the components down the hierarchy (in this case **OptionItem**).
+
+The following case shows a list of cards (with a title and image inside). We need to change the card appearance using `provide()` and `inject()`. Those cards have four styles: rounded, squared borders, and dark and light themes. This pattern is ideal for implementing more cards along the same behavior.
+
+{% img "4-visualization-preferences.png" "Provide/Inject design pattern example, vue patterns" "lazy" %}
+
+So, to implement this feature, we need to create the image structure in the code, beginning with `OptionItem.vue`.
+
+```ts
+    // File: OptionItem.vue
+    
+    <template>
+      <div
+        class="option"
+        :class="[`option--${boders}`, darkMode && 'option--dark']"
+      >
+        <div :class="['option__title', darkMode && 'option__title--dark']">
+          {{ title }}
+        </div>
+        <div class="option__image">
+          <img :src="image" :alt="title" />
+        </div>
+      </div>
+    </template>
+    
+    <script lang="ts">
+    import { defineComponent, inject } from "vue";
+    import { VISUALIZATION_PREFERENCES } from "./OptionSymbols";
+    
+    export interface OptionProps {
+      title: string;
+      image: string;
+    }
+    
+    export default defineComponent({
+      props: {
+        title: {
+          type: String,
+          default: undefined,
+        },
+        image: {
+          type: String,
+          default: undefined,
+        },
+      },
+      setup() {
+        // The important part 👇
+        const injection = inject(VISUALIZATION_PREFERENCES);
+    
+        return {
+          darkMode: injection?.darkMode,
+          boders: injection?.borders,
+        };
+      },
+    });
+    </script>
+    
+    <style scoped>
+    /* styles doesn't matter for now */
+    </style>
+```
+
+`OptionItem.vue` pretends to be a card with a title and a random image inside.
+
+{% img "5-card.png" "OptionItem.vue Component" "lazy" %}
+
+The `setup()` uses the `inject()` method to retrieve (inject) the data from the `provider()` method, to be used in `OptionContainer.vue` (two components above in the hierarchy). 
+
+To inject that data inside the component, we need a key. In this case, `VISUALIZATION\_PREFERENCES` is an `InjectionKey`. Its use makes sense with Typescript. It allows you to define a key name and structure for a specific data structure. The `VISUALIZATION\_PREFERENCES` is in the `OptionSymbols.ts` file for re-usability purposes.
```

Comment:
I think slashes \ are not needed in both  `VISUALIZATION\_PREFERENCES`

Reply:
modified and done, thanks